salt/modules/glusterfs.py

```diff
@@ -43,34 +43,46 @@ def _get_minor_version():
     return version
 
 
-def _gluster(cmd):
+def _gluster_ok(xml_data):
     '''
-    Perform a gluster command.
+    Extract boolean return value from Gluster's XML output.
     '''
-    # We will pass the command string as stdin to allow for much longer
-    # command strings. This is especially useful for creating large volumes
-    # where the list of bricks exceeds 128 characters.
-    return __salt__['cmd.run'](
-        'gluster --mode=script', stdin="{0}\n".format(cmd))
+    return int(xml_data.find('opRet').text) == 0
 
 
 def _gluster_xml(cmd):
     '''
-    Perform a gluster --xml command and check for and raise errors.
+    Perform a gluster --xml command and log result.
     '''
+    # We will pass the command string as stdin to allow for much longer
+    # command strings. This is especially useful for creating large volumes
+    # where the list of bricks exceeds 128 characters.
     root = ET.fromstring(
         __salt__['cmd.run'](
             'gluster --xml --mode=script', stdin="{0}\n".format(cmd)
         ).replace("\n", ""))
-    if int(root.find('opRet').text) != 0:
-        raise CommandExecutionError(root.find('opErrstr').text)
+    if _gluster_ok(root):
+        output = root.find('output')
+        if output:
+            log.info('Gluster call "{0}" succeeded: {1}'.format(cmd, root.find('output').text))
+        else:
+            log.info('Gluster call "{0}" succeeded'.format(cmd))
+    else:
+        log.error('Failed gluster call: {0}: {1}'.format(cmd, root.find('opErrstr').text))
     return root
```

Comment:
Yes, they can raise errors and the code still raises some (salt invocation errors). A good question is, which types of errors should do that? Just every error or only a subset like bad invocations. I've no answers or even strong opinions about this. In addition to the internal use in states, module functions are directly callable by users and I was thinking that nice return values would be appreciated there. Although, I don't know how Salt reports exceptions in that case. I found the current behaviour of immediately raising an exception for a failing Gluster call problematic as it produced some bugs. For instance, a failing status() call raised an unhandled exception while the calling code was supposed to do things when an error occurs. Of course, this could've been fixed by improving exception handling. My modifications may loose some verbosity (excluding logs) but this could be fixed by returning dictionaries with both boolean error and output string. At least, I hope that after my modifications the code is easily improvable despite of which decisions are made on this specific question.